### PR TITLE
[MU4] Xml reader/writer

### PR DESCRIPTION
**Pull request by RomanPudashkin**
Implemented XmlReader/Writer as part of mu4 framework

### CODE REVIEW

File: src/framework/global/xmlreader.h [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+    std::string attribute(std::string_view name) const;
+    bool hasAttribute(std::string_view name) const;
+
+    enum ReadStringBehaviour {
+        ErrorOnUnexpectedElement,
+        IncludeChildElements,
+        SkipChildElements
+    };
+
+    std::string readString(ReadStringBehaviour behaviour = ErrorOnUnexpectedElement);
```
Comment (Eism) [inline src/framework/global/xmlreader.h:69]: can we replace these methods with one method Val read(behavior)?

File: src/framework/global/xmlreader.h [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+
+    TokenType readNext();
+    TokenType tokenType() const;
+
+    int intAttribute(std::string_view name, int defaultValue = 0) const;
+    double doubleAttribute(std::string_view name, double defaultValue = 0.) const;
+    std::string attribute(std::string_view name) const;
+    bool hasAttribute(std::string_view name) const;
+
+    enum ReadStringBehaviour {
```
Comment (Eism) [inline src/framework/global/xmlreader.h:63]: typo: Behavior

File: src/framework/global/xmlreader.h [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+        StartElement,
+        EndElement,
+        Characters,
+        Comment
+    };
+
+    TokenType readNext();
+    TokenType tokenType() const;
+
+    int intAttribute(std::string_view name, int defaultValue = 0) const;
```
Comment (Eism) [inline src/framework/global/xmlreader.h:58]: same
Val attribute(std::string_view name, const Val& defaultValue = Val())

File: src/framework/midi/internal/midiconfiguration.cpp [hunk 4]
```diff
@@ -131,97 +130,79 @@ io::path MidiConfiguration::stateFilePath() const
...
 
     xml.writeEndElement();
     xml.writeEndDocument();
 
     if (xml.hasError()) {
         LOGE() << "failed write xml";
-        return false;
-    }
-
-    if (!file.flush()) {
```
Comment (Eism) [inline src/framework/midi/internal/midiconfiguration.cpp:221]: should we flush the file in the XmlWriter?

File: src/framework/workspace/iworkspacedatastreamregister.h [hunk 5]
```diff
@@ -19,22 +19,20 @@
...
 class IWorkspaceDataStreamRegister : MODULE_EXPORT_INTERFACE
 {
     INTERFACE_ID(IWorkspaceDataStreamRegister)
+
 public:
     virtual ~IWorkspaceDataStreamRegister() = default;
 
-    virtual void regStream(const std::string& tag, std::shared_ptr<IWorkspaceDataStream> stream) = 0;
-    virtual std::shared_ptr<IWorkspaceDataStream> stream(const std::string& tag) const = 0;
+    virtual void addStream(IWorkspaceDataStreamPtr stream) = 0;
```
Comment (Eism) [inline src/framework/workspace/iworkspacedatastreamregister.h:33]: why renamed? :)

File: src/framework/global/xmlreader.h (hunk 2 above)
Comment (RomanPudashkin) [reply src/framework/global/xmlreader.h]: fixed, thanks

File: src/framework/global/xmlreader.h (hunk 1 above)
Comment (RomanPudashkin) [reply src/framework/global/xmlreader.h]: Discussed that issues with Elnur, decided left as is

File: src/framework/workspace/iworkspacedatastreamregister.h (hunk 5 above)
Comment (RomanPudashkin) [reply src/framework/workspace/iworkspacedatastreamregister.h]: Returned old name

File: src/framework/midi/internal/midiconfiguration.cpp (hunk 4 above)
Comment (RomanPudashkin) [reply src/framework/midi/internal/midiconfiguration.cpp:221]: QFile does flush() in close()
https://doc.qt.io/qt-5/qfiledevice.html#close

**Review by Eism — APPROVED** [review]

**Review by igorkorsukov — APPROVED** [review]

**Pull request merged**